In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section
of the German eBay website
We sampled 50,000 data points from the full dataset.The aim of this project is to clean the data and analyze the included used car listings. 

# 2.1.7.1 Introduction

To open the csv file by panda

In [1]:
import pandas as pd
import numpy as np
autos= pd.read_csv('ebay.csv',encoding='Windows-1252')


To explore the csv file

In [2]:
autos
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# observation of the above code
 1. dataset contains 20 columns, most of which are strings.
 2. Some columns have null values.
 3. The column names use camelcase instead of Python's preferred snakecase.

# 2.1.7.2 cleaning the column name

1. at first the name of the 'yearOfRegistration','monthOfRegistration','notRepairedDamage','dateCreated' is converted directy by rewriting and assigning the column name
2. to convert the camelcase to snakecase a function name 'cameltosnake' is created
3. using for loop we converted the camelcase into snakecase into the col_list

In [3]:
col_name=(autos.columns)
col_name

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
col_name=['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen']

def cameltosnake (string):
    output= [string[0].lower()]
    for char in string[1:]:
        if ord(char) in range(65,90):
            output.append('_')
            output.append(char.lower())
        else :
            output.append(char)
    return str.join('',output)

for ele in col_name:
    col_name[col_name.index(ele)]= cameltosnake(ele)

autos.columns= col_name
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# 2.1.7.3 initial exploration and cleaning

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

1. To Remove  ','   and   'km'   character from 'odometer' column
2. To change 'odometer' datatype from sting to int
3. To change column name 'odometer' to 'odometer_km'

In [6]:
autos['odometer']= autos['odometer'].str.replace(',','')
autos['odometer']= autos['odometer'].str.replace('km','')
autos['odometer']= autos['odometer'].astype(int)
autos.rename(columns={'odometer':'odometer_km'},inplace=True)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type    price   abtest vehicle_type  registration_year  \
0      privat    Angebot   $5,000  control          bus               2004   
1      privat    Angebot   $8,500  control    limousine               1997   
2      privat    Angebot   $8,990     test    limousine               2009   
3      privat    Angebot   $4,350  control   kleinwagen               2007   
4      privat    Angebot   $1,350     test        kombi               2003   
...       ...        ...      ...      ...          ...                ...   
49995  privat    Angebot  $24,900  control    limousine               2011   
49996  privat    Angebot   $1,980  control       cabrio               1996   
49997  privat    Angebot  $13,200     test       cabrio               2014   
49998  privat    Angebot  $22,900  control        kombi               2013   
49999  privat    Angebot   $1,250  control    limousine               1996   

         gearbox  power_p_s   model  odometer_km  registration_month  \
0        manuell        158  andere       150000                   3   
1      automatik        286     7er       150000                   6   
2        manuell        102    golf        70000                   7   
3      automatik         71  fortwo        70000                   6   
4        manuell          0   focus       150000                   7   
...          ...        ...     ...          ...                 ...   
49995  automatik        239      q5       100000                   1   
49996    manuell         75   astra       150000                   5   
49997  automatik         69     500         5000                  11   
49998    manuell        150      a3        40000                  11   
49999    manuell        101  vectra       150000                   1   

      fuel_type       brand unrepaired_damage           ad_created  \
0           lpg     peugeot              nein  2016-03-26 00:00:00   
1        benzin         bmw              nein  2016-04-04 00:00:00   
2        benzin  volkswagen              nein  2016-03-26 00:00:00   
3        benzin       smart              nein  2016-03-12 00:00:00   
4        benzin        ford              nein  2016-04-01 00:00:00   
...         ...         ...               ...                  ...   
49995    diesel        audi              nein  2016-03-27 00:00:00   
49996    benzin        opel              nein  2016-03-28 00:00:00   
49997    benzin        fiat              nein  2016-04-02 00:00:00   
49998    diesel        audi              nein  2016-03-08 00:00:00   
49999    benzin        opel              nein  2016-03-13 00:00:00   

       nr_of_pictures  postal_code            last_seen  
0                   0        79588  2016-04-06 06:45:54  
1                   0        71034  2016-04-06 14:45:08  
2                   0        35394  2016-04-06 20:15:37  
3                   0        33729  2016-03-15 03:16:28  
4                   0        39218  2016-04-01 14:38:50  
...               ...          ...                

1. To Remove ',' and 'km' character from 'price' column
2. To change 'price' column datatype from string to int

In [7]:
autos['price']= autos['price'].str.replace('$','')
autos['price']= autos['price'].str.replace(',','')
autos['price']= autos['price'].astype(int)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_p_s   model  odometer_km  registration_month  \
0        manuell        158  andere       150000                   3   
1      automatik        286     7er       150000                   6   
2        manuell        102    golf        70000                   7   
3      automatik         71  fortwo        70000                   6   
4        manuell          0   focus       150000                   7   
...          ...        ...     ...          ...                 ...   
49995  automatik        239      q5       100000                   1   
49996    manuell         75   astra       150000                   5   
49997  automatik         69     500         5000                  11   
49998    manuell        150      a3        40000                  11   
49999    manuell        101  vectra       150000                   1   

      fuel_type       brand unrepaired_damage           ad_created  \
0           lpg     peugeot              nein  2016-03-26 00:00:00   
1        benzin         bmw              nein  2016-04-04 00:00:00   
2        benzin  volkswagen              nein  2016-03-26 00:00:00   
3        benzin       smart              nein  2016-03-12 00:00:00   
4        benzin        ford              nein  2016-04-01 00:00:00   
...         ...         ...               ...                  ...   
49995    diesel        audi              nein  2016-03-27 00:00:00   
49996    benzin        opel              nein  2016-03-28 00:00:00   
49997    benzin        fiat              nein  2016-04-02 00:00:00   
49998    diesel        audi              nein  2016-03-08 00:00:00   
49999    benzin        opel              nein  2016-03-13 00:00:00   

       nr_of_pictures  postal_code            last_seen  
0                   0        79588  2016-04-06 06:45:54  
1                   0        71034  2016-04-06 14:45:08  
2                   0        35394  2016-04-06 20:15:37  
3                   0        33729  2016-03-15 03:16:28  
4                   0        39218  2016-04-01 14:38:50  
...               ...          ...                  ...  
49995           

# observation
1. (seller and offer_type) has almost the same value of 4999 times in 5000 entry.
2. there are others columns like(abtest, gearbox, unrepaired_damage) contains only two unique value.


# 2.1.7.4 exploring the odometer and price column 


In [8]:
print(autos['odometer_km'].unique())
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].value_counts().sort_index())
print(autos['odometer_km'].head())

[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]
(13,)
5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32


In [9]:
print(autos['price'].unique())
print(autos['price'].unique().shape)
print(autos['price'].value_counts().sort_index().head(40))

[ 5000  8500  8990 ...   385 22200 16995]
(2357,)
0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
Name: price, dtype: int64


In [10]:
del_col=autos['price']<1000
del_rows= autos[del_col].index
autos.drop(del_rows, inplace=True)
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 38640, dtype: int32

observation:
    The contents of 'odometer_km' column seems accurate but the content of 'price' column seems inaccurate as there exist
    minimum_value=0. so it is inaccurate.
    as a solution we drop the rows in which the price of a car is less than $1000 which is appropriate as the minimum value of 
    a car. 

# 2.1.7.5. exploring the date column

In [12]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna= False).sort_index(ascending=True)

2016-03-05    0.025543
2016-03-06    0.013872
2016-03-07    0.035119
2016-03-08    0.032660
2016-03-09    0.032479
2016-03-10    0.033307
2016-03-11    0.032790
2016-03-12    0.037396
2016-03-13    0.015994
2016-03-14    0.036620
2016-03-15    0.033618
2016-03-16    0.029063
2016-03-17    0.030487
2016-03-18    0.012836
2016-03-19    0.035119
2016-03-20    0.038147
2016-03-21    0.037319
2016-03-22    0.032557
2016-03-23    0.032195
2016-03-24    0.029011
2016-03-25    0.030512
2016-03-26    0.033100
2016-03-27    0.031392
2016-03-28    0.035352
2016-03-29    0.034006
2016-03-30    0.033049
2016-03-31    0.031418
2016-04-01    0.034601
2016-04-02    0.036284
2016-04-03    0.039130
2016-04-04    0.036905
2016-04-05    0.013354
2016-04-06    0.003261
2016-04-07    0.001501
Name: date_crawled, dtype: float64

At 2016-04-03, highest ad was first crawled where the percentage is 0.039130 

In [13]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna= False).sort_index(ascending=True)

2015-06-11    0.000026
2015-08-10    0.000026
2015-09-09    0.000026
2015-11-10    0.000026
2015-12-30    0.000026
                ...   
2016-04-03    0.039441
2016-04-04    0.037319
2016-04-05    0.011982
2016-04-06    0.003364
2016-04-07    0.001320
Name: ad_created, Length: 74, dtype: float64

In [14]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna= False).sort_index(ascending=True)

2016-03-05    0.001087
2016-03-06    0.003571
2016-03-07    0.004555
2016-03-08    0.006263
2016-03-09    0.008929
2016-03-10    0.009808
2016-03-11    0.011724
2016-03-12    0.022205
2016-03-13    0.008385
2016-03-14    0.011982
2016-03-15    0.014984
2016-03-16    0.015450
2016-03-17    0.026372
2016-03-18    0.007376
2016-03-19    0.014596
2016-03-20    0.019798
2016-03-21    0.019695
2016-03-22    0.020782
2016-03-23    0.017909
2016-03-24    0.018530
2016-03-25    0.017754
2016-03-26    0.016071
2016-03-27    0.014079
2016-03-28    0.019436
2016-03-29    0.020782
2016-03-30    0.023447
2016-03-31    0.022748
2016-04-01    0.023188
2016-04-02    0.024922
2016-04-03    0.024431
2016-04-04    0.023370
2016-04-05    0.131134
2016-04-06    0.234705
2016-04-07    0.139933
Name: last_seen, dtype: float64

In [15]:
autos['registration_year'].describe()

count    38640.000000
mean      2005.676423
std         86.670173
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

1. minimum value is 1000 when cars were not invented
2. maximum value is 9999 which is a invalid year indicating the future.

# 2.1.7.6. dealing with incorrect registration year of data

In [28]:
valid_reg_year= autos['registration_year'] > 2016
autos.loc[valid_reg_year,'registration_year'].describe


<bound method NDFrame.describe of 68       2017
113      2017
164      2018
253      2017
348      2017
         ... 
49689    2017
49696    2017
49731    2017
49796    2017
49910    9000
Name: registration_year, Length: 1420, dtype: int64>

In [29]:
valid_reg_year= autos['registration_year']<1900
autos.loc[valid_reg_year,'registration_year'].describe

<bound method NDFrame.describe of 22316    1000
49283    1001
Name: registration_year, dtype: int64>

observation:
    total number of rows that is not in (1900-2016) is 1421. As the number of records is a handsome amount so applying custom 
    logic is preffered.